### ISSUE TITLE
Lynx v0.3.0 & v0.3.1 makes illegal move

### ISSUE THREAD

**Issue by tissatussa** [lynx-chess/Lynx#31]
while testing in CuteChess, your latest Lynx versions 0.3.0 and 0.3.1 both made an illegal move early in a game and lost .. this happened all 3 times i tested, so it must be a severe bug.  I have Xubuntu 20.04 and i used your Linux built version from the Assets.

here are the games :

```
[Event "?"]
[Site "?"]
[Date "2021.09.16"]
[Round "?"]
[White "Lynx v0.3.0"]
[Black "MinimalChessEngine v0.6.1"]
[Result "0-1"]
[ECO "A05"]
[GameDuration "00:02:20"]
[Opening "Reti Opening"]
[PlyCount "6"]
[Termination "illegal move"]
[TimeControl "600+2"]

1. Nf3 {+0.20/8 20s} Nf6 {-0.02/17 43s} 2. d4 {+0.20/8 20s} d5 {-0.25/16 8.6s}
3. Nc3 {+0.10/7 11s} e6 {+0.14/15 12s, White makes an illegal move: f1g2} 0-1

[Event "?"]
[Site "?"]
[Date "2021.09.16"]
[Round "?"]
[White "Chess22K v1.14"]
[Black "Lynx v0.3.0"]
[Result "1-0"]
[FEN "r1bqk2r/pppp1Npp/2n2n2/4p3/2B1P3/8/PPPP1bPP/RNBQK2R w KQkq - 0 6"]
[GameDuration "00:02:28"]
[PlyCount "7"]
[SetUp "1"]
[Termination "illegal move"]
[TimeControl "600+2"]

1. Kf1 {+1.94/24 23s} Qe7 {-4.70/7 20s} 2. Nxh8 {+1.69/23 5.2s}
Bd4 {-4.55/6 14s} 3. d3 {+2.23/21 28s} d6 {-4.45/6 20s}
4. Nd2 {+2.37/23 12s, Black makes an illegal move: e8c8} 1-0

[Event "?"]
[Site "?"]
[Date "2021.09.17"]
[Round "?"]
[White "Alouette v0.1.5"]
[Black "Lynx v0.3.1"]
[Result "1-0"]
[ECO "A45"]
[GameDuration "00:02:13"]
[Opening "Queen's pawn game"]
[PlyCount "13"]
[Termination "illegal move"]
[TimeControl "600+2"]

1. d4 Nf6 {-0.20/7 20s} 2. Nc3 e6 {-0.15/7 20s} 3. Bg5 c5 {+0.10/7 19s} 4. dxc5
Bxc5 {+0.20/7 19s} 5. Nf3 Nc6 {+0.25/6 19s} 6. e4 Qb6 {+0.75/6 18s}
7. Qd2 {Black makes an illegal move: b2a1} 1-0
```

**Comment by eduherminio** [comment]
Hi @tissatussa, thanks a lot for the report.

Would you be so kind to provide the log files of the engine?
There should be inside a `/logs` dir at the executable level: `logs/logs-2021-09-17.log`, `logs/errors-2021-09-17.log`, etc.

**Comment by tissatussa** [comment]
here they are :
[logs_v0.3.0.zip](https://github.com/lynx-chess/Lynx/files/7185043/logs_v0.3.0.zip)
[logs_v0.3.1.zip](https://github.com/lynx-chess/Lynx/files/7185045/logs_v0.3.1.zip)

**Comment by eduherminio** [comment]
There is indeed something wrong with my engine, more specifically about how it handles `stop` commands.

However, CuteChess is also doing something weird in those games that I haven't seen before: it gives the engine a `position` command, makes it `stop`, changes their opponent's last move and then gives the engine the new `position` command.
I haven't been able to reproduce that behavior in CuteChess Windows version.

Out of curiosity, are you using it in any fancy way that could explain such behavior?

Examples:
```
2021-09-16 21:00:01.4187|[GUI]	position fen r1bqk2r/pppp1Npp/2n2n2/4p3/2B1P3/8/PPPP1bPP/RNBQK2R w KQkq - 0 6 moves e1f1 d8e7 f7h8 f2d4 d2d3 d7d6 c1g5 
2021-09-16 21:00:01.4187|[GUI]	isready 
2021-09-16 21:00:01.4187|[Lynx]	readyok 
2021-09-16 21:00:01.4187|[GUI]	go ponder wtime 552598 btime 553813 winc 2000 binc 2000 
2021-09-16 21:00:13.8540|[GUI]	stop  
2021-09-16 21:00:20.6763|[Lynx]	info depth 6 seldepth 6 multipv 1 score cp -415 nodes 3132893 nps 154741 time 19246 pv c8e6 c4e6 e7e6 g5f6 e6f6 
2021-09-16 21:00:20.6763|[Lynx]	bestmove c8e6 ponder c4e6 
2021-09-16 21:00:20.6763|[GUI]	position fen r1bqk2r/pppp1Npp/2n2n2/4p3/2B1P3/8/PPPP1bPP/RNBQK2R w KQkq - 0 6 moves e1f1 d8e7 f7h8 f2d4 d2d3 d7d6 b1d2 // b1d2 instead of c1g5
2021-09-16 21:00:20.6763|[GUI]	isready 
2021-09-16 21:00:20.6763|[Lynx]	readyok
2021-09-16 21:00:20.6763|[GUI]	go wtime 542170 btime 553813 winc 2000 binc 2000 
```

```
2021-09-17 10:56:12.2271|[GUI]	position startpos moves d2d4 g8f6 b1c3 e7e6 c1g5 c7c5 d4c5 f8c5 g1f3 b8c6 e2e4 d8b6 e4e5 
2021-09-17 10:56:12.2271|[GUI]	isready 
2021-09-17 10:56:12.2271|[Lynx]	readyok 
2021-09-17 10:56:12.2271|[GUI]	go ponder wtime 613736 btime 498981 winc 2000 binc 2000 
2021-09-17 10:56:12.3550|[GUI]	stop 
2021-09-17 10:56:12.3867|[Lynx]	info depth 4 seldepth 4 multipv 1 score cp 75 nodes 26263 nps 22699 time 157 pv b6b2 e5f6 b2c3 
2021-09-17 10:56:12.3867|[Lynx]	bestmove b6b2 ponder e5f6 
2021-09-17 10:56:12.3867|[GUI]	position startpos moves d2d4 g8f6 b1c3 e7e6 c1g5 c7c5 d4c5 f8c5 g1f3 b8c6 e2e4 d8b6 d1d2 // d1d2 instead of e4e5
2021-09-17 10:56:12.3867|[GUI]	isready 
2021-09-17 10:56:12.3867|[Lynx]	readyok 
2021-09-17 10:56:12.3867|[GUI]	go wtime 615610 btime 498981 winc 2000 binc 2000 
```

**Comment by tissatussa** [comment]
i'm using the latest CuteChess v1.2.0, no Beta version or so .. and i make no fancy settings, just a starting FEN and i often set a bonus time, eg. 3 seconds .. i have no problems with most engines in CuteChess .. which test can i do to help ?

**Comment by eduherminio** [comment]
No worries, it was out of curiosity, since I didn't experience this issue in my local env.
A fix is already on its way, will ask you to test a preview version before releasing it as part of v0.3.2.

**Comment by eduherminio** [comment]
@tissatussa Would you mind giving [v0.3.2-preview.1](https://github.com/lynx-chess/Lynx/releases/tag/v0.3.2-preview.1) a try 🙏🏼?

**Comment by tissatussa** [comment]
this version seems to work fine !
i'm only able to try the linux-x64 version, because i use no Windows or Mac.
i set the Hash size from 1 (!? =low) to 64 (=normal) but the debug log says something like "Hash size modification not supported yet" .. nevermind, but then you could leave that Hash option out (for now) .. i also changed the Depth option from 3 to 9 and it works, however i don't see any difference ..

in one game Lynx had only 1 legal move (at a certain moment, when it was in check), but Lynx was thinking for several seconds !? A few other engines have this issue, eg. Weiss had it and i wrote the author and he changed it.

**Comment by eduherminio** [comment]
> this version seems to work fine !

That's great to hear, will release a non-preview version by the end of the day,

> i set the Hash size from 1 (!? =low) to 64 (=normal) but the debug log says something like "Hash size modification not supported yet" .. nevermind, but then you could leave that Hash option out (for now) .. i also changed the Depth option from 3 to 9 and it works, however i don't see any difference ..

I'm still working in making `hash` size and `depth` command work properly.
TBH, `depth` should currently show the same message as `hash` does, since it's not limiting search depth for now.
Or even better, as you suggested, just not show the engine option until it's ready. Will do the latter very soon.

> in one game Lynx had only 1 legal move (at a certain moment, when it was in check), but Lynx was thinking for several seconds !? A few other engines have this issue, eg. Weiss had it and i wrote the author and he changed it.

I have that in my TODO list as well, but thanks for pointing out!

---

In general I'm tracking new features/improvements privately (Trello board) instead of using GitHub issues, mostly because:
* I wasn't really expecting anyone to follow or test my engine, at least initially
* There are tons of things still pending to be implemented, but since this is my first chess engine, I'm still learning about most of them!

I might reconsider this approach in the near future and start moving tickets here, but meanwhile feel free to log any other issues/suggestions/comments.

Thanks for taking the time to test Lynx and for these comments!

**Comment by tissatussa** [comment]
i'm glad to help .. i test several chess engines i find on GitHub and i try to compile them .. that way i learn a lot, because authors use many different languages to build their engine, eg. C, D, go, rust, nim, and more .. sometimes i have to use a pre-compiled binary, such as your Lynx .. can it be compiled ?  You supplied no Makefile ..

Btw. I tested Lynx playing White and Black, and i also gave a FEN starting position .. all seem to work fine.

BONUS - today i discovered this position : r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R w KQkq - 0 1
On the TalkChess forum they wrote :
_the above position is from "kiwipete" (Peter McKenzie). It has a little bit of everything so it's a good position to use for perft._
Maybe i can help you while developing :-)

**Comment by eduherminio** [comment]
> BONUS - today i discovered this position : r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R w KQkq - 0 1
On the TalkChess forum they wrote :
the above position is from "kiwipete" (Peter McKenzie). It has a little bit of everything so it's a good position to use for perft.
Maybe i can help you while developing :-)

Thanks, I'm already using it in some places (see [this search](https://github.com/lynx-chess/Lynx/search?q=r3k2r%2Fp1ppqpb1%2Fbn2pnp1%2F3PN3%2F1p2P3%2F2N2Q1p%2FPPPBBPPP%2FR3K2R+w+KQkq)).

> can it be compiled ? You supplied no Makefile ..

Lynx can definitely be compiled, in fact, CI system does it for every commit I push.
But one needs to be familiar with .NET environment (or check [CI build script](https://github.com/lynx-chess/Lynx/blob/main/.github/workflows/ci.yml), which is obv not ideal, my bad 😓.

I'll work on adding a Makefile that does it for you together with some instructions in the README, since Makefiles seem to be 'universal' enough!

**Comment by eduherminio** [comment]
Fixed in [v0.4.0](https://github.com/lynx-chess/Lynx/releases/tag/v0.4.0) (there were other changes already merged, so I had to bump the minor version).
